### ISSUE TITLE
Warn destructing attrs & slots

### ISSUE THREAD

**Issue by jd-solanki** [vuejs/eslint-plugin-vue#1964]
**Please describe what the rule should do:**

**What category should the rule belong to?**

[ ] Enforces code style (layout)
[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```js
setup(props, { attrs, slots }) {
  const { slotFoo, slotBar } = slots
  const { id, hidden } = attrs
}
```

**Additional context**
According to vue [docs](https://vuejs.org/api/composition-api-setup.html#setup-context), destructured `attrs` & `slots` aren't reactive. Hence, we should not destructure them in `setup` function.

> attrs and slots are stateful objects that are always updated when the component itself is updated. This means you should avoid destructuring them and always reference properties as attrs.x or slots.x. Also note that, unlike props, the properties of attrs and slots are not reactive.

EDIT: Sample code updated, Thanks @FloEdelmann

**Comment by FloEdelmann** [comment]
I think your example code is fine. But this would not be:

```js
setup(props, { attrs, slots }) {
  const { slotFoo, slotBar } = slots
  const { id, hidden } = attrs
}
```

**Comment by jd-solanki** [comment]
Oh, Yes. I totally misunderstood that.

docs are pointing to not destructure the `attrs` & `slots`. I thought it destructing the ctx. Silly me.

Do we have eslint rule for it.

I will update the sample code.

**Comment by FloEdelmann** [comment]
We have a similar rule: https://eslint.vuejs.org/rules/no-setup-props-destructure.html

Maybe it could be modified to also check attrs and slots.

**Comment by FloEdelmann** [comment]
Related to #1949.

**Comment by ota-meshi** [comment]
Hmm...
I don't think `attrs` and `slots` are reactive, destructured or not.

So they should be used more carefully. I find it difficult to check well.

For example, use onBeforeUpdate to handle:

```js
setup(props, { attrs, slots }) {
  let { slotFoo, slotBar } = slots
  let { id, hidden } = attrs
  // ...
  onBeforeUpdate(() => {
    ;({ slotFoo, slotBar } = slots)
    ;({ id, hidden } = attrs)
    // ...
  })
}
```

Using destructuring, but I don't think the above is wrong.

**Comment by jd-solanki** [comment]
> I don't think `attrs` and `slots` are reactive

They are stateful as mentioned in [docs](https://vuejs.org/api/composition-api-setup.html#setup-context)

**Comment by ota-meshi** [comment]
Yes. In other words, the example code you provided is neither right nor wrong.
Even with destructuring, it will probably behave correctly depending on how the value of is handled.

```js
setup(props, { attrs, slots }) {
   const { slotFoo, slotBar } = slots
   const { id, hidden } = attrs
}
```


What would your proposed rule check for?

**Comment by jd-solanki** [comment]
> What would your proposed rule check for?

It will assure that we don't destructure the attrs & slots.

In my lib [anu-vue](https://anu-vue.netlify.app/) I previously had [this](https://github.com/jd-solanki/anu/blob/87fcf75054/packages/anu-vue/src/components/base-input/ABaseInput.tsx#L33) destructuring which was causing reactivity lost.

E.g. If I pass `type` attribute as ref and change it via button click then its `type` value doesn't get updated from `text` to `password`.

Thanks.

**Comment by ota-meshi** [comment]
Did you fix it? I think the `attrs.id` of the `main` branch is still lost in reactivity.

https://github.com/jd-solanki/anu/blob/6796c843812cb9ec3cd1e0cd337866df40967439/packages/anu-vue/src/components/base-input/ABaseInput.tsx#L24

**Comment by jd-solanki** [comment]
> I think the attrs.id of the main branch is still lost in reactivity

Let me check

**Comment by jd-solanki** [comment]
Below code in my playground is working fine on main branch:
```vue
<script lang="ts" setup>
import { ref } from "vue";

const inputType = ref("password");
const toggleType = () => {
  console.log("clicked");
  inputType.value = inputType.value === "text" ? "password" : "text";
};
</script>

<template>
    <AInput :type="inputType" modelValue="text"></AInput>
    <ABtn @click="toggleType">Toggle type</ABtn>
</template>
```

It is working fine:
![chrome-capture-2022-8-12](https://user-images.githubusercontent.com/47495003/189614038-0452c782-da22-4252-9431-4d2fba013ae6.gif)

Now even docs are confusing me 🤣 

It states attrs & slots "stateful objects". Moreover, there's also "properties of attrs and slots are **not** reactive".

IG someone better than me needs to comment on this or I should try this in SFC playground.

**Comment by ota-meshi** [comment]
Can you provide a SFC playground link to try out your components?
